service/history/workflow_util.go

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) 2017-2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"time"
+
+	"github.com/pborman/uuid"
+
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/types"
+	"github.com/uber/cadence/service/history/execution"
+	"github.com/uber/cadence/service/history/shard"
+)
+
+type (
+	UpdateWorkflowAction struct {
+		Noop           bool
+		CreateDecision bool
+	}
+
+	UpdateWorkflowActionFunc func(execution.Context, execution.MutableState) (*UpdateWorkflowAction, error)
+)
+
+var (
+	UpdateWorkflowWithNewDecision = &UpdateWorkflowAction{
+		CreateDecision: true,
+	}
+	UpdateWorkflowWithoutDecision = &UpdateWorkflowAction{
+		CreateDecision: false,
+	}
+)
+
+///////////////////  Util function for loading workflow ///////////////////
+
+func LoadWorkflowOnce(
```

Comment:
are those functions cover in the handler tests?

Reply:
Good point. I think right now those util functions are covered by `history_engine_test.go` as when calling history engine APIs, those functions will also be called. But there's no specific unit test for those util functions. 

When moving those functions to a separate `workflow` package, I think it make sense to add those specific tests. I will do that in the diff for separating packages.